From ca9759063c1cecbcb8f683026aa57183696b9ede Mon Sep 17 00:00:00 2001
From: Pardha Kesapragada <pkesapragada@habana.ai>
Date: Mon, 22 Feb 2021 13:14:59 +0200
Subject: [PATCH] [SW-20154] gaudi2: change the efuse bit definitions

eFUSE bit definitions for PCIe recovery code location,
keys placement have been changed by the security
acrhitecture today, after a discussion. Hence this
patch makes corresponding changes.

Change-Id: I84ccc683ae2eb460730e9e61dbb4299dc6ead103
Signed-off-by: Pardha Kesapragada <pkesapragada@habana.ai>
---

diff --git a/gaudi2_arc_gnu/common.c b/gaudi2_arc_gnu/common.c
index 3775721..28c8ef0 100644
--- a/gaudi2_arc_gnu/common.c
+++ b/gaudi2_arc_gnu/common.c
@@ -273,9 +273,19 @@
 
 void execute_pcie_recovery(void)
 {
-	/* if eFUSE == ROM_PCIE_Recovery, execute directly */
-	if (arcboot_rom_global.config_fuse &
-			EFUSE_SYSCFG_RECOVERY_CODE_LOCATION_ROM) {
+	/*
+	 * efuse[14:13]:
+	 * 00: ROM
+	 * 01: eROM
+	 * 10: Reserved
+	 * 11: ROM
+	 */
+	uint8_t pcie_location = ((arcboot_rom_global.config_fuse >>
+			EFUSE_SYSCFG_RECOVERY_CODE_LOCATION) &
+			EFUSE_SYSCFG_RECOVERY_CODE_LOCATION_MASK);
+
+	if (pcie_location != EFUSE_SYSCFG_RECOVERY_CODE_LOCATION_EROM) {
+		/* In all other scenarios, execute from ROM (most secure) */
 		pcie_recovery();
 		return;
 	}
diff --git a/gaudi2_arc_gnu/dev_efuse.h b/gaudi2_arc_gnu/dev_efuse.h
index 9f5728e..57550a9 100644
--- a/gaudi2_arc_gnu/dev_efuse.h
+++ b/gaudi2_arc_gnu/dev_efuse.h
@@ -4,8 +4,14 @@
  */
 
 #define EFUSE_KEY_HASH_OFFSET				0x14
+
+/*
+ * 00: ROM, 01: EROM, 10: RSVD, 11:ROM
+ */
 #define EFUSE_SYSCFG_RECOVERY_CODE_LOCATION		13
-#define EFUSE_SYSCFG_RECOVERY_CODE_LOCATION_MASK	0x1
+#define EFUSE_SYSCFG_RECOVERY_CODE_LOCATION_MASK	0x3
+#define EFUSE_SYSCFG_RECOVERY_CODE_LOCATION_EROM	0x1
+
 #define EFUSE_SYSCFG_ENABLE_TPM_MASK			BIT(11)
 #define EFUSE_SYSCFG_FW_VERSION_LOCATION		7
 #define EFUSE_SYSCFG_FW_VERSION_LOCATION_MASK		0x3
@@ -15,12 +21,11 @@
 #define EFUSE_SYSCFG_REVOKATION_LOCATION_MASK		0x3
 
 #define EFUSE_SYSCFG_FW_VERSION_LOCATION_SPI		BIT(8)
-#define EFUSE_SYSCFG_KEY_LOCATION_ROM			BIT(5)
 #define EFUSE_SYSCFG_REVOKATION_LOCATION_SPI		BIT(8)
-#define EFUSE_SYSCFG_RECOVERY_CODE_LOCATION_ROM		BIT(13)
 
-#define EFUSE_SYSCFG_DEFAULT	(EFUSE_SYSCFG_SPI_CONTROLLER | \
-				 EFUSE_SYSCFG_ENABLE_VALIDATION_MASK | \
-				 EFUSE_SYSCFG_FW_VERSION_LOCATION_SPI | \
-				 EFUSE_SYSCFG_KEY_LOCATION_ROM | \
-				 EFUSE_SYSCFG_REVOKATION_LOCATION_SPI)
+/*
+ * As per recommendation from Security Architect, if FUSE is not available,
+ * card should not continue in the secure boot mode. Hence the default
+ * can be 0 (with no enabling of the Signature of Validation also.
+ */
+#define EFUSE_SYSCFG_DEFAULT				0
diff --git a/gaudi2_arc_gnu/dev_security.h b/gaudi2_arc_gnu/dev_security.h
index 64bc68b..9247604 100644
--- a/gaudi2_arc_gnu/dev_security.h
+++ b/gaudi2_arc_gnu/dev_security.h
@@ -17,8 +17,8 @@
 #define NUM_DWORDS_AVLBL_IN_PSOC_SEC			8
 
 enum security_key_location {
-	KEY_LOCATION_EFUSE,
-	KEY_LOCATION_ROM
+	KEY_LOCATION_ROM,
+	KEY_LOCATION_EFUSE
 };
 
 /* size of key properties in bytes */
